pkg/vm/stackitem/json.go

```diff
@@ -260,72 +261,110 @@ func (d *decoder) decodeMap() (*Map, error) {
+	if item == nil {
+		return nil, fmt.Errorf("%w: nil", ErrUnserializable)
+	}
 	if len(seen) > MaxJSONDepth {
-		return "", ErrTooDeep
+		return nil, ErrTooDeep
+	}
+	if old, ok := seen[item]; ok {
+		if old.end == 0 {
+			// Compound item marshaling which has not yet finished.
+			return nil, ErrRecursive
+		}
+		if len(data)+old.end-old.start > MaxSize {
+			return nil, errTooBigSize
+		}
+		return append(data, data[old.start:old.end]...), nil
+	}
+
+	if _, ok := item.(Null); ok {
+		if len(data)+14 > MaxSize {
+			return nil, errTooBigSize
+		}
+		data = append(data, `{"type":"Any"}`...)
+		return data, nil
+	}
+
+	start := len(data)
+
+	prefix := `{"type":"` + item.Type().String() + `","value":`
+	if len(data)+len(prefix) > MaxSize {
+		return nil, errTooBigSize
 	}
-	var value interface{}
+	data = append(data, prefix...)
+
+	// Primitive stack items are appended after the switch
+	// to reduce the amount of size checks.
+	var primitive string
+	var err error
+
 	switch it := item.(type) {
 	case *Array, *Struct:
-		if seen[item] {
-			return "", ErrRecursive
-		}
-		seen[item] = true
-		arr := []interface{}{}
-		for _, elem := range it.Value().([]Item) {
-			s, err := toJSONWithTypes(elem, seen)
+		seen[item] = sliceNoPointer{start, 0}
+		data = append(data, '[')
+		for i, elem := range it.Value().([]Item) {
+			if i != 0 {
+				data = append(data, ',')
+			}
+			data, err = toJSONWithTypes(data, elem, seen)
 			if err != nil {
-				return "", err
+				return nil, err
 			}
-			arr = append(arr, s)
 		}
-		value = arr
-		delete(seen, item)
 	case Bool:
-		value = bool(it)
+		if it {
+			primitive = "true"
+		} else {
+			primitive = "false"
+		}
 	case *Buffer, *ByteArray:
-		value = base64.StdEncoding.EncodeToString(it.Value().([]byte))
+		primitive = `"` + base64.StdEncoding.EncodeToString(it.Value().([]byte)) + `"`
 	case *BigInteger:
-		value = it.Big().String()
+		primitive = `"` + it.Big().String() + `"`
 	case *Map:
-		if seen[item] {
-			return "", ErrRecursive
-		}
-		seen[item] = true
-		arr := []interface{}{}
+		seen[item] = sliceNoPointer{start, 0}
+		data = append(data, '[')
 		for i := range it.value {
-			// map keys are primitive types and can always be converted to json
-			key, _ := toJSONWithTypes(it.value[i].Key, seen)
-			val, err := toJSONWithTypes(it.value[i].Value, seen)
+			if i != 0 {
+				data = append(data, ',')
+			}
+			data = append(data, `{"key":`...)
+			data, err = toJSONWithTypes(data, it.value[i].Key, seen)
+			if err != nil {
+				return nil, err
+			}
+			data = append(data, `,"value":`...)
+			data, err = toJSONWithTypes(data, it.value[i].Value, seen)
 			if err != nil {
-				return "", err
+				return nil, err
 			}
-			arr = append(arr, map[string]interface{}{
-				"key":   key,
-				"value": val,
-			})
+			data = append(data, '}')
 		}
-		value = arr
-		delete(seen, item)
 	case *Pointer:
-		value = it.pos
-	case nil:
```

Comment:
You've removed `nil`, but we also have `InteropInterface` type.

Reply:
Fixed.